### ISSUE TITLE
Unable to run pyinstaller on USB key (Other drive than C:)

### ISSUE THREAD

**Issue by ghrlt** [pyinstaller/pyinstaller#6590]
## Description of the issue

Pyinstaller crash with this `OSError: [WinError 123] La syntaxe du nom de fichier, de répertoire ou de volume est incorrecte: 'E:\\C:'`

### Context information (for bug reports)
Trying to build an exe from a Python script located on my usb key

* Output of `pyinstaller --version`: ```4.7```
* Version of Python:  3.10.1
* Platform:  Windows 10 
* How you installed Python:  python.org
* Did you also try this on another platform? Does it work there? No

* try the latest development version, using the following command:

```shell
pip install https://github.com/pyinstaller/pyinstaller/archive/develop.zip
```

* follow *all* the instructions in our "If Things Go Wrong" Guide
  (https://github.com/pyinstaller/pyinstaller/wiki/If-Things-Go-Wrong) and

### Make sure [everything is packaged correctly](https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs#make-sure-everything-is-packaged-correctly)

  * [x] start with clean installation
  * [ ] use the latest development version
  * [ ] Run your frozen program **from a command window (shell)** — instead of double-clicking on it
  * [ ] Package your program in **--onedir mode**
  * [ ] Package **without UPX**, say: use the option `--noupx` or set `upx=False` in your .spec-file
  * [ ] Repackage you application in **verbose/debug mode**. For this, pass the option `--debug` to `pyi-makespec` or `pyinstaller` or use `EXE(..., debug=1, ...)` in your .spec file.

### A minimal example program which shows the error
Any program, but located on another drive than C:
```
(paste text here)
“Minimal“ means: remove everything from your code which is not relevant for this bug,
esp. don't use external programs, remote requests, etc.
A very good example is https://gist.github.com/ronen/024cdae9ff2d50488438. This one helped
us reproducing and fixing a quite complex problem within approx 1 hour.
```

### Stacktrace / full error message
```
E:\$temp\Python\HItori\Hitori>pyinstaller --onefile --noconsole app.py
731 INFO: PyInstaller: 4.7
732 INFO: Python: 3.10.1
795 INFO: Platform: Windows-10-10.0.19042-SP0
808 INFO: wrote E:\$temp\Python\HItori\Hitori\app.spec
855 INFO: UPX is not available.
Traceback (most recent call last):
  File "C:\Users\XX\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\XX\AppData\Local\Programs\Python\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\XX\AppData\Local\Programs\Python\Python310\Scripts\pyinstaller.exe\__main__.py", line 7, in <module>
  File "C:\Users\XX\AppData\Local\Programs\Python\Python310\lib\site-packages\PyInstaller\__main__.py", line 124, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "C:\Users\XX\AppData\Local\Programs\Python\Python310\lib\site-packages\PyInstaller\__main__.py", line 58, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "C:\Users\XX\AppData\Local\Programs\Python\Python310\lib\site-packages\PyInstaller\building\build_main.py", line 782, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "C:\Users\XX\AppData\Local\Programs\Python\Python310\lib\site-packages\PyInstaller\building\build_main.py", line 670, in build
    os.makedirs(pth, exist_ok=True)
  File "C:\Users\XX\AppData\Local\Programs\Python\Python310\lib\os.py", line 215, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "C:\Users\XX\AppData\Local\Programs\Python\Python310\lib\os.py", line 215, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "C:\Users\XX\AppData\Local\Programs\Python\Python310\lib\os.py", line 215, in makedirs
    makedirs(head, exist_ok=exist_ok)
  [Previous line repeated 6 more times]
  File "C:\Users\XX\AppData\Local\Programs\Python\Python310\lib\os.py", line 225, in makedirs
    mkdir(name, mode)
OSError: [WinError 123] La syntaxe du nom de fichier, de répertoire ou de volume est incorrecte: 'E:\\C:'
```

Please also see <https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs>
for more about what would use to solve the issue.

**Comment by bwoodsend** [comment]
I develop across drives by default so I know that this isn't true generally,

> 808 INFO: wrote E:\\$temp\Python\HItori\Hitori\app.spec

Can you share the output of `set` from cmd? And your current working directory? That `$temp` shouldn't be there.

**Comment by ghrlt** [comment]
Yes I think it is an issue with my path folders name, anyway, it shouldn't
be a problem
Had others issues in others software with this.

**Comment by ghrlt** [comment]
Here is output of `set`
```
E:\$temp\Python\HItori\Hitori>set
ALLUSERSPROFILE=C:\ProgramData
APPDATA=C:\Users\XX\AppData\Roaming
CommonProgramFiles=C:\Program Files\Common Files
CommonProgramFiles(x86)=C:\Program Files (x86)\Common Files
CommonProgramW6432=C:\Program Files\Common Files
COMPUTERNAME=COMPUTER-NAME
ComSpec=C:\windows\system32\cmd.exe
DriverData=C:\Windows\System32\Drivers\DriverData
FPS_BROWSER_APP_PROFILE_STRING=Internet Explorer
FPS_BROWSER_USER_PROFILE_STRING=Default
HOMEDRIVE=C:
HOMEPATH=\Users\XX
LOCALAPPDATA=C:\Users\XX\AppData\Local
LOGONSERVER=\\COMPUTER-NAME
NUMBER_OF_PROCESSORS=4
OneDrive=C:\Users\XX\OneDrive
OS=Windows_NT
Path=C:\Program Files (x86)\Common Files\Oracle\Java\javapath;C:\ProgramData\Oracle\Java\javapath;C:\windows\system32;C:\windows;C:\windows\System32\Wbem;C:\windows\System32\WindowsPowerShell\v1.0\;C:\windows\System32\OpenSSH\;C:\Program Files\Intel\TXE Components\DAL\;C:\Program Files (x86)\Intel\TXE Components\DAL\;C:\Program Files\Git\cmd;C:\platform-tools;C:\Users\XX\AppData\Local\Programs\Python\Python310\Scripts\;C:\Users\XX\AppData\Local\Programs\Python\Python310\;C:\Users\XX\AppData\Local\Microsoft\WindowsApps;
PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC
PROCESSOR_ARCHITECTURE=AMD64
PROCESSOR_IDENTIFIER=Intel64 Family 6 Model 122 Stepping 8, GenuineIntel
PROCESSOR_LEVEL=6
PROCESSOR_REVISION=7a08
ProgramData=C:\ProgramData
ProgramFiles=C:\Program Files
ProgramFiles(x86)=C:\Program Files (x86)
ProgramW6432=C:\Program Files
PROMPT=$P$G
PSModulePath=C:\Program Files\WindowsPowerShell\Modules;C:\windows\system32\WindowsPowerShell\v1.0\Modules
PUBLIC=C:\Users\Public
SESSIONNAME=Console
SystemDrive=C:
SystemRoot=C:\windows
TEMP=C:\Users\XX\AppData\Local\Temp
TMP=C:\Users\XX\AppData\Local\Temp
USERDOMAIN=COMPUTER-NAME
USERDOMAIN_ROAMINGPROFILE=COMPUTER-NAME
USERNAME=XX
USERPROFILE=C:\Users\XX
windir=C:\windows
ZES_ENABLE_SYSMAN=1
```

**Comment by bwoodsend** [comment]
I think I see the chain of events now. Your current working directory (which contains `$temp`) is included in `workpath` and `distpath` which then [get normalised](https://github.com/pyinstaller/pyinstaller/blob/v4.7/PyInstaller/building/build_main.py#L635-L636). That normalisation [includes evaluating `$xxx` environment variables](https://github.com/pyinstaller/pyinstaller/blob/v4.7/PyInstaller/compat.py#L567) so they become `E:\C:\Users\XX\AppData\Local\Temp\Python\HItori\Hitori\build` and `E:\C:\Users\XX\AppData\Local\Temp\Python\HItori\Hitori\dist`. Then those folders [are passed to `os.mkdirs()`](https://github.com/pyinstaller/pyinstaller/blob/v4.7/PyInstaller/building/build_main.py#L670) which recursively tries to create parent directories and eventually failing on the root folder `E:\C:`.

So TL;DR don't use special characters in file names.

**Comment by ghrlt** [comment]
Yep, gonna have to replace the $ by some _ I guess
Anyway, thanks you for your explanations.
